libbeat/processors/processor.go

```diff
@@ -35,11 +36,26 @@ type Processors struct {
 	log  *logp.Logger
 }
 
+// Processor is the interface that all processors must implement
 type Processor interface {
 	Run(event *beat.Event) (*beat.Event, error)
 	String() string
 }
 
+// Closer defines the interface for processors that should be closed after using
+// them.
+type Closer interface {
+	Close() error
+}
+
+// Close closes a processor if it implements the Closer interface
+func Close(p Processor) error {
+	if closer, ok := p.(Closer); ok {
```

Comment:
I know this makes the interface for Processors (and this PR) smaller . But I would prefer to not have dynamic semi-hidden features by casting into alternative interfaces. How about adding `Close` to the Processor interface directly?

Alternatively to adding `Close` we could pass a TaskGroup to the processor constructor (processor factory). This would provide a cancellation context and waiting mechanism to processors that want to start tasks in the background without having the need to add `Close`.

Reply:
Ok, I will give it a try.

Reply:
Ok, I tried some things, but I feel slightly more comfortable following the current approach.

> How about adding Close to the Processor interface directly?

This sounds good to me, no problem with making the PR bigger :slightly_smiling_face: But we would lose the chance of using this to distinguish processors that need cleanup in the script processor: https://github.com/elastic/beats/pull/16349/files#diff-41e65bb8d42b3e093252d491f0011957R155
Would you have other suggestions to make this check? An option is not to do it, we wouldn't be worse than we are now, and we could leave this for a future change.

> Alternatively to adding Close we could pass a TaskGroup to the processor constructor (processor factory). This would provide a cancellation context and waiting mechanism to processors that want to start tasks in the background without having the need to add Close.

I have also given a chance to this approach, first quick and dirty results can be found here: https://github.com/elastic/beats/pull/21720/files
This works for some cases I have manually tested, but I would need to test it a lot more, and many things would still need to be improved, and tests rewritten. It also has the problem of changing some interfaces.
I would prefer not to embark on this effort now.

Reply:
Added a comment about why the processor and closer interfaces are separated now.